Add vault secrets guide (#7653)
solo-io/gloo: docs/content/installation/advanced_configuration/vault_secrets.md

---
title: Storing Gloo Edge secrets in HashiCorp Vault
weight: 50
description: Using HashiCorp Vault as a backing store for Gloo Edge secrets
---

Use [HashiCorp Vault Key-Value storage](https://www.vaultproject.io/docs/secrets/kv/kv-v2.html) as a backing store for Gloo Edge secrets.

> When Gloo Edge boots, it reads a {{< protobuf name="gloo.solo.io.Settings">}} resource from a preconfigured location. By default, Gloo Edge attempts to read a `gloo.solo.io/v1.Settings` Custom Resource named `default` in the `gloo-system` namespace of your Kubernetes cluster. By editing this settings file, you can configure Vault as the secret store for your Edge environment.

Reviewer: do you know what exactly is stored in Vault or this Kube secret? I think it would be nice to say something like this: 

When Gloo Edge boots, it reads the {{< protobuf name="gloo.solo.io.Settings">}} custom resource named `default` in the `gloo-system` namespace of your Kubernetes cluster to receive the proxy configuration for the gateway. By default, this configuration references a Kubernetes secret that Gloo Edge uses to ???? (whatever is in there and what it is used for). If you want to store the ???? (whatever is in there) in a Vault instance instead of a Kubernetes secret, you can edit the Settings custom resource to point to this Vault secret store so that Gloo Edge retrieves the ??? from Vault.

Author: These fields `kubernetesSecretSource` or `directorySecretSource` or `vaultSecretSource` are not specifying where to look for a secret at startup, rather they specify where to read/write all secrets throughout Gloo Edge's operations. `kubernetesSecretSource` doesn't require any additional configuration because Gloo Edge knows how to connect to the cluster in which it is running by default.

{{% notice tip %}}
Want to use Vault with Gloo Edge outside of Kubernetes instead? You can provide your settings file to Gloo Edge inside of a configuration directory when you [run Gloo Edge locally]({{< versioned_link_path fromRoot="/installation/gateway/development/docker-compose-file">}}).
{{% /notice %}}

## Customizing the Gloo Edge settings file

> Edit the `default` settings resource so Gloo Edge reads and writes secrets using HashiCorp Vault.

Reviewer: maybe we should add that Vault needs to be set up separately either in the cluster or externally, just to make sure that this is not covered as part of this guide.

Reviewer 2: would the vault instance need to be in the same cluster for this setup to work as we have it written? like if the vault instance is elsewhere, would we need to have extra steps for how to connect to that instance first

Author: It would not need to be in the same cluster, it just needs to be routable on the network at the location provided in the `address` field

1. Edit the `default` settings resource.
   ```shell script
   kubectl --namespace gloo-system edit settings default
   ```

2. Make the following changes to the resource.

>    * Remove the existing `kubernetesSecretSource` or `directorySecretSource` field, which is required for the Vault secret storage to be used.

Reviewer: ```suggestion
   * Remove the existing `kubernetesSecretSource` or `directorySecretSource` field.
```

Author: These fields `kubernetesSecretSource` or `directorySecretSource` or `vaultSecretSource` are not specifying where to look for a secret at startup, rather they specify where to read/write all secrets throughout Gloo Edge's operations. `kubernetesSecretSource` doesn't require any additional configuration because Gloo Edge knows how to connect to the cluster in which it is running by default.

   * Add the `vaultSecretSource` section to enable secrets to be read from and written to Vault.

>    * Add the `refreshRate` field, which is used for watching Vault secrets and the local filesystem for changes.

Reviewer: is local filesystem  related to when Vault runs on your local machine? or is this something else?

Reviewer: question: do people need to create Vault secrets before gloo edge is booted? if so, we should add this as a step before this one

Author: Not necessarily related to where Vault is run, rather where Gloo Edge is run. The secrets do not need to be created before Edge is booted; it will poll Vault for changes

   {{< highlight yaml "hl_lines=16-25" >}}
   apiVersion: gloo.solo.io/v1
   kind: Settings
   metadata:
     name: default
     namespace: gloo-system
   spec:
     discoveryNamespace: gloo-system
     gateway:
       validation:
         alwaysAccept: true
         proxyValidationServerAddr: gloo:9988
     gloo:
       xdsBindAddr: 0.0.0.0:9977
     kubernetesArtifactSource: {}
     kubernetesConfigSource: {}
     # Delete or comment out the existing 
     # kubernetesSecretSource or directorySecretSource field
     #kubernetesSecretSource: {}
     # Enable secrets to be read from and written to HashiCorp Vault
     vaultSecretSource:
       address: http://vault:8200
       token: root
     # refresh rate for polling config backends for changes

>      # this is used for watching vault secrets and the local filesystem

Author: ```suggestion
     # this is used for watching vault secrets and by other resource clients
```

## Writing secret objects to Vault

After configuring Vault as your secret store, be sure to write any Vault secrets by using Gloo Edge-style YAML. You can either use the `glooctl create secret` command or manually write secrets.

### Using glooctl

To get started writing Gloo Edge secrets for use with Vault, you can use the `glooctl create secret` command. A benefit of using `glooctl` for secret creation is that the secret is created in the path that Gloo Edge watches.

For example, you might use the following command to create a secret in Vault.
```bash
glooctl create secret tls \
    --certchain /path/to/cert.pem \
    --privatekey /path/to/key.pem
    --name tls-secret \
    --use-vault \
    --vault-address http://vault:8200/ \
    --vault-token "$VAULT_TOKEN"
```
This command creates a TLS secret with the following value:
```json
{
  "metadata": {
    "name": "tls-secret",
    "namespace": "gloo-system"
  },
  "tls": {
    "certChain": "-----BEGIN CERTIFICATE-----\n...\n-----END CERTIFICATE-----\n",
    "privateKey": "-----BEGIN PRIVATE KEY-----\n...\n-----END PRIVATE KEY-----\n"
  }
}

> ```

Reviewer: wondering if there is a way to tell that the secret creation in Vault worked? is there maybe a status on the secret that you can check?

Author: This is a good question; I will check on this

Author: There would not be a way to check a status on the secret to my knowledge since we don't have k8s resources for them when they're stored in Vault

You can also include the `-o json` flag in the command for JSON-formatted secrets, which can be manually stored as values in Vault.

### Manually writing secrets

> Be sure to write any Vault secrets by using Gloo Edge-style YAML. For more information, see the {{< protobuf name="gloo.solo.io.Secret" display="v1.Secret API reference">}}.

Reviewer: I did not understand at first that this requires you to create a secret as well. maybe say something like: 

Instead of using the `glooctl create secret` command to create the Vault secret and automatically store the secret key in your Vault instance, you can use your own configuration file to create the secret. Note that you must use the same YAML format for the secret so that  Gloo Edge can read the secret.

If you manually write Gloo Edge secrets, you must store them in Vault with the correct Vault key names, which adhere to the following format:

`<secret_engine_path_prefix>/<gloo_root_key>/<resource_group>/<group_version>/Secret/<resource_namespace>/<resource_name>`

| Path | Description |
| ---- | ----------- |

> | `<secret_engine_path_prefix>` | The `pathPrefix` configured in the Settings `vaultSecretSource`. Defaults to `secret`. Note that the default path for the kv secrets engine in Vault is `kv` when Vault is not run with `-dev`. |

Reviewer: this pathPrefix is not in our example above I think. Since this key is so complicated, maybe we could just show an example Vault secret key here?

Reviewer: I think it would be something like: 

`secret/root/gloo.solo.io/v1/gloo-system/my-secret`

| `<gloo_root_key>` | The `rootKey` configured in the Settings `vaultSecretSource`. Defaults to `gloo` |

> | `<resource_group>` | The API group/proto package in which resources of the given type are contained. {{< protobuf name="gloo.solo.io.Secret" display="Gloo Edge secrets">}} have the resource group `gloo.solo.io`. |

Reviewer: ```suggestion
| `<resource_group>` | The API group/proto package in which resources of the given type are contained. The {{< protobuf name="gloo.solo.io.Secret" display="Gloo Edge secrets">}} custom resource has the resource group `gloo.solo.io`. |
```
